rfcs/0000-occupied-capacity-increment-rate/0000-occupied-capacity-increment-rate.md

```diff
@@ -0,0 +1,73 @@
+---
+Number: 0000
+Category: Standards Track
+Status: Proposal
+Author: Jinyang Jiang
+Organization: Nervos Foundation
+Created: 2019-2-11
+---
+
+# Occupied capacity increment rate
+
+## Abstract
+
+Capacity is one of the core resources in CKB network, itself is just a number representing how much storage resource a user can use.
+
+Once a user decides to store something in CKB, the user must lock some capacity in exchange, the locked capacity which we called occupied capacity represent the actually used resource in the network.
+
+Occupied capacity affects CKB network in two ways:
+
+1. Single block occupied capacity: if a single block or transaction contains too much occupied capacity, it may affect the propagation of the block and may slow down the whole network.
+2. The whole network occupied capacity: that every full node must cost the same memory or disk to save these content, it incentive miners more centralize which is not making any sense for a decentralized network.
+
+So the solution is to limit occupied capacity in each block, to indirectly control the network's total occupied capacity and the increment rate of occupied capacity.
+
+## Specification
+
+Define two variables:
+
+* `MAX_BLOCK_BYTES`                 - the max size of a block, value: 10 MB
+* `MAX_INCREMENT_OCCUPIED_CAPACITY` - the max increment occupied capacity of a block, value: 2 MB
+
+Add two verification rules to CKB consensus:
+
+### 1. Block size verification rule
+
+Serialized block size must smaller or equal than `MAX_BLOCK_BYTES`.
+
+Pseudo code: 
+``` rust
+assert!(serialized_block_size <= MAX_BLOCK_BYTES)
+```
+
+CKB use [CFB][1] as the serializer, consider the serializer formats should not affect consensus rule, the `serialized_block_size` should be the actual byte size of data that do not include metadata introduced by CFB.
```

Comment:
```suggestion
CKB uses [CFB][1] as the serializer. Considering that the serializer format should not affect consensus rule, the `serialized_block_size` should be the actual byte size of data excluding metadata introduced by CFB.
```

Reply:
It is unclear what is the metadata introduced by CFB.

Reply:
Flatbuffer encodes extra information into serialized binary to describe the struct(for example length of a list). So I assume CFB will do the same thing.

Which I mean is block size consensus rule should not consider that extra information (metadata); instead, we should verify the actual size(remove padding and metadata which CFB introduced) of a block.

Reply:
Bitcoin uses serialized size in consensus rule, but in our case, that CFB is a memory efficient serialization library and it is more complicated than bitcoin's serialization.
It is mean it will be hard to upgrade CFB format if the consensus rule depends on CFB serialization.